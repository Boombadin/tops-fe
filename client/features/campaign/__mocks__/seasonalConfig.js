export default {
  sku: ['123', '456', '789'],
  remark: {
    th: {
      homepage: 'หมายเหตุหน้าโฮม',
    },
    en: {
      homepage: 'homepage remark',
    },
  },
  style: {
    web: {
      border_color: '#f70405',
      homepage_remark_font_color: '#f70405',
      banner: {
        head_title_header_icon_url: 'https://i.ibb.co/0V4yL9f/group-11-3x.png',
        tail_title_header_icon_url:
          'https://i.ibb.co/h8cwn5n/china-lamp-right-3x.png',
        text_link_font_color: '#FFFFFF',
        text_link_url: 'https://www.google.com',
        title_header_font_color: '#ffdb71',
        th: {
          banner_image_url: 'https://i.ibb.co/2Y4jMKS/f40405.png',
          text_link: 'แสดงทั้งหมด',
          title_header: 'เทศกาลวันสาร์สจีน',
        },
        en: {
          banner_image_url: 'https://i.ibb.co/2Y4jMKS/f40405.png',
          text_link: 'Show All',
          title_header: 'Chinese Ghost Festival',
        },
      },
    },
  },
};
