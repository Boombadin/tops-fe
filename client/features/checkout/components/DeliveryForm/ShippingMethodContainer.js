import { breakpoint } from '@central-tech/core-ui';
import { find, get, isEmpty } from 'lodash';
import moment from 'moment';
import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import styled, { css } from 'styled-components';

import { ButtonRadio } from '@client/components/ButtonRadio';
import { createIntervals } from '@client/components/TimeSlotTab/utils';
import { TextGuide } from '@client/components/Typography';
import TimeSlotContainer from '@client/features/checkout/components/DeliveryForm/TimeSlotContainer';
import { getShippingMethod } from '@client/features/checkout/utils';
import withLocales from '@client/hoc/withLocales';
import {
  ELEMENT_ACTION,
  ELEMENT_TYPE,
  generateTestId,
} from '@client/utils/generateElementId';
import { format } from '@client/utils/time';
const CALENDAR_ICON = '/assets/icons/calendar-icon.svg';
const GRAB_ICON = '/assets/images/Grab_logo.svg';

const Section = styled.div`
  width: 100%;
  height: auto;
  border-bottom: dashed 1px #e8e8e8;
  padding: 10px 20px;
  ${breakpoint('xs', 'md')`
    border-bottom: solid 1px #e9e9e9;
  `}
`;

const SectionTitle = styled.div`
  height: 60px;
  width: 900px;
  display: -ms-flexbox;
  display: table-cell;
  -webkit-align-items: center;
  vertical-align: middle;
  -ms-flex-align: center;
`;

const SectionTitleIcon = styled.img`
  height: 25px;
  margin: 0 0 -6px;
`;

const SectionContent = styled.div`
  width: 100%;
  height: auto;

  ${props => props.padding && `padding: ${props.padding};`}
  
  ${props =>
    props.inline &&
    `
      display: flex;
    `}
  ${props =>
    props.column &&
    css`
      display: flex;
      flex-direction: column;
      ${breakpoint('md')`
        flex-direction: row;
      `}
    `}
`;

const ShippingMethod = styled.div`
  display: flex;
  align-items: center;
  width: 100%;
  height: 100px;
  border-radius: 4px;
  border: solid 1px #e8e8e8;
  background-color: #fafafa;
  padding: 10px;
  margin-top: 10px;
  cursor: pointer;

  ${breakpoint('md')`
    width: 380px;
    margin-right: 20px;

    @media (max-width: 1169px) {
      height: auto;
    }
  `}

  ${props =>
    props.active &&
    `
      background-color: rgba(128, 189, 0, 0.1);
      border: solid 1px #80bd00;
    `}
`;

const ShippingMethodDescription = styled.div`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const ShippingMethodDescriptionTitle = styled.div`
  font-size: 14px;
  font-weight: bold;
  color: #2a2a2a;
  display: flex;
  align-items: center;
  height: 28px;

  ${breakpoint('md')`
    @media (max-width: 979px) {
      display: block;
      height: auto;
    }
  `}
`;

const MethodTitle = styled.span``;

const MethodPowered = styled.div`
  display: flex;
  align-items: center;
  height: 28px;
`;

const ShippingMethodDescriptionTitleLabel = styled(TextGuide)`
  color: #666666;
  padding: 0 10px;

  ${breakpoint('md')`
    @media (max-width: 979px) {
      font-weight: bold;
      padding: 0;
    }
  `}
`;

const ShippingMethodDescriptionText = styled(TextGuide)`
  color: #666666;
`;

const GrabIcon = styled.img`
  height: 22px;
  margin-bottom: 5px;
`;

const NoSlotAvalible = styled.span`
  font-size: 18px;
  color: red;
`;

class ShippingMethodContainer extends PureComponent {
  state = {
    shippingMethod: null,
    currentSlotId: '',
  };

  componentDidUpdate(prevProps) {
    if (
      !isEmpty(this.props.shippingMethods) &&
      prevProps.shippingMethods !== this.props.shippingMethods
    ) {
      this.checkDefaultShippingMethod();
    }
  }

  componentDidMount() {
    this.checkDefaultShippingMethod();
    const slotId = `${get(
      this.props,
      'cart.extension_attributes.shipping_date',
    )}/${get(this.props, 'cart.extension_attributes.shipping_slot_id')}`;
    this.setState({ currentSlotId: slotId });
  }

  checkDefaultShippingMethod = () => {
    if (this.props.deliveryMethod === 'home_delivery') {
      const methodCode = get(
        find(this.props.shippingMethods, x => x.carrierCode === 'standard'),
        'methodCode',
        'mds',
      );
      const standard = getShippingMethod(
        this.props.shippingMethods,
        methodCode,
      );
      const express = getShippingMethod(this.props.shippingMethods, 'express');
      this.setState({
        shippingMethod: express?.nextRound
          ? express?.method?.methodCode
          : standard?.method?.methodCode,
      });
    }
  };

  handleShippingMethod = method => {
    this.setState({
      shippingMethod: method,
    });
    this.props.onShippingChange(method);
  };

  render() {
    const { shippingMethod, currentSlotId } = this.state;
    const {
      shippingAddress,
      shippingMethods,
      storeLocator,
      storeLocatorSlot,
      onChooseSlot,
      translate,
      lang,
      deliveryMethod,
      collector,
      cart,
      clearShipping,
    } = this.props;

    let shippingMethodSelected;
    if (!isEmpty(shippingMethod)) {
      shippingMethodSelected = get(
        find(shippingMethods, { methodCode: shippingMethod }),
        'extensionAttributes.deliverySlot',
        [],
      );
    } else {
      shippingMethodSelected = get(
        find(shippingMethods, { carrierCode: 'standard' }),
        'extensionAttributes.deliverySlot',
        [],
      );
    }
    // Back Up to Next Version
    // let intervals = [];
    // const defaultDays = 7;
    // if (deliveryMethod === 'click_and_collect') {
    //   intervals = createIntervals(
    //     storeLocatorSlot,
    //     storeLocatorSlot?.length > defaultDays
    //       ? storeLocatorSlot?.length
    //       : defaultDays,
    //   );
    // } else {
    //   intervals = createIntervals(
    //     shippingMethodSelected,
    //     shippingMethodSelected?.length > defaultDays ||
    //       shippingMethod === 'express'
    //       ? shippingMethodSelected?.length
    //       : defaultDays,
    //   );
    // }

    let intervals = [];
    const defaultDays = 7;
    if (deliveryMethod === 'click_and_collect') {
      // !isEmpty(storeLocatorSlot) && intervals.push({ days: storeLocatorSlot });
      intervals = createIntervals(
        storeLocatorSlot,
        storeLocatorSlot?.length > defaultDays
          ? defaultDays
          : storeLocatorSlot?.length,
      );
    } else {
      intervals = createIntervals(
        shippingMethodSelected,
        shippingMethodSelected?.length > defaultDays
          ? defaultDays
          : shippingMethodSelected?.length,
      );
    }
    let mobileIntervals = [];
    if (deliveryMethod === 'click_and_collect') {
      !isEmpty(storeLocatorSlot) &&
        mobileIntervals.push({ days: storeLocatorSlot });
    } else {
      mobileIntervals = createIntervals(
        shippingMethodSelected,
        shippingMethodSelected?.length,
      );
    }
    const methodCode = get(
      find(shippingMethods, x => x.carrierCode === 'standard'),
      'methodCode',
      'mds',
    );
    const standard = getShippingMethod(shippingMethods, methodCode);
    const express = getShippingMethod(shippingMethods, 'express');
    return (
      <React.Fragment>
        <Section>
          <SectionTitle
            data-testid={generateTestId({
              type: ELEMENT_TYPE.INFO,
              action: ELEMENT_ACTION.VIEW,
              moduleName: 'ShippingMethodContainer',
              uniqueId: 'ShippingTitle',
            })}
          >
            <TextGuide type="topic" bold>
              <SectionTitleIcon
                src={CALENDAR_ICON}
                style={{ marginRight: 10 }}
              />
              {deliveryMethod === 'home_delivery'
                ? translate(
                    'checkout_delivery.shipping_method.home_delivery.title',
                  )
                : translate(
                    'checkout_delivery.shipping_method.click_and_collect.title',
                  )}
            </TextGuide>
          </SectionTitle>
          {deliveryMethod &&
            deliveryMethod === 'home_delivery' &&
            !clearShipping && (
              <SectionContent column>
                {/* Express */}
                {!isEmpty(get(express, 'nextRound')) && (
                  <ShippingMethod
                    data-testid={generateTestId({
                      type: ELEMENT_TYPE.BUTTON,
                      action: ELEMENT_ACTION.CHANGE,
                      moduleName: 'ShippingMethodContainer',
                      uniqueId: 'SelectShippingExpress',
                    })}
                    onClick={() =>
                      this.handleShippingMethod(
                        get(express, 'method.methodCode', 'express'),
                      )
                    }
                    active={
                      shippingMethod ===
                      get(express, 'method.methodCode', 'express')
                        ? 'active'
                        : ''
                    }
                  >
                    <ButtonRadio
                      name="shippingMethod"
                      checked={
                        shippingMethod ===
                        get(express, 'method.methodCode', 'express')
                          ? 'checked'
                          : ''
                      }
                      readOnly
                    />

                    <ShippingMethodDescription>
                      <ShippingMethodDescriptionTitle>
                        <MethodTitle>
                          {translate(
                            'checkout_delivery.shipping_method.shipping_title_method.express',
                          )}
                        </MethodTitle>
                        <MethodPowered>
                          <ShippingMethodDescriptionTitleLabel type="caption-2">
                            {translate(
                              'right_menu.shipping_options.powered_by',
                            )}
                          </ShippingMethodDescriptionTitleLabel>
                          <GrabIcon src={GRAB_ICON} />
                        </MethodPowered>
                      </ShippingMethodDescriptionTitle>

                      <TextGuide color="#666666" type="caption-2">
                        {express?.slotLabel ||
                          translate(
                            'checkout_delivery.shipping_method.express.head_subtext',
                          )}
                        <br />
                        {translate(
                          'right_menu.shipping_options.next_round',
                        )}{' '}
                        {translate(
                          'right_menu.shipping_options.next_round_datetime',
                          {
                            date:
                              moment(moment(), 'YYYY-MM-DD', lang) ===
                              get(express, 'nextRound.date')
                                ? translate('timeslot.grid.today')
                                : format(
                                    get(express, 'nextRound.date'),
                                    lang === 'en_US' ? 'ddd' : 'dddd',
                                    lang,
                                  ),
                            time: `${get(
                              express,
                              'nextRound.slot.timeFrom',
                            )} - ${get(express, 'nextRound.slot.timeTo')}`,
                          },
                        )}
                        <br />
                        {translate('right_menu.shipping_options.round_cost', {
                          deliveryFee: get(express, 'nextRound.slot.cost', '-'),
                        })}
                      </TextGuide>
                    </ShippingMethodDescription>
                  </ShippingMethod>
                )}

                {/* Standard */}
                {!isEmpty(get(standard, 'nextRound')) && (
                  <ShippingMethod
                    data-testid={generateTestId({
                      type: ELEMENT_TYPE.BUTTON,
                      action: ELEMENT_ACTION.CHANGE,
                      moduleName: 'ShippingMethodContainer',
                      uniqueId: 'SelectShippingStandard',
                    })}
                    onClick={() =>
                      this.handleShippingMethod(
                        get(standard, 'method.methodCode', 'mds'),
                      )
                    }
                    active={
                      shippingMethod ===
                      get(standard, 'method.methodCode', 'mds')
                        ? 'active'
                        : ''
                    }
                  >
                    <ButtonRadio
                      name="shippingMethod"
                      checked={
                        shippingMethod ===
                        get(standard, 'method.methodCode', 'mds')
                          ? 'checked'
                          : ''
                      }
                      readOnly
                    />
                    <ShippingMethodDescription>
                      <ShippingMethodDescriptionTitle>
                        {translate(
                          'checkout_delivery.shipping_method.shipping_title_method.delivery',
                        )}
                      </ShippingMethodDescriptionTitle>
                      <ShippingMethodDescriptionText type="caption-2">
                        {standard?.slotLabel ||
                          translate(
                            'checkout_delivery.shipping_method.standard.head_subtext',
                            {
                              day: get(
                                standard,
                                'method.extensionAttributes.deliverySlot',
                                [],
                              ).length,
                            },
                          )}
                        <br />
                        {translate(
                          'right_menu.shipping_options.next_round',
                        )}{' '}
                        {translate(
                          'right_menu.shipping_options.next_round_datetime',
                          {
                            date:
                              moment(moment(), 'YYYY-MM-DD', lang) ===
                              get(standard, 'nextRound.date')
                                ? translate('timeslot.grid.today')
                                : format(
                                    get(standard, 'nextRound.date'),
                                    lang === 'en_US' ? 'ddd' : 'dddd',
                                    lang,
                                  ),
                            time: `${get(
                              standard,
                              'nextRound.slot.timeFrom',
                            )} - ${get(standard, 'nextRound.slot.timeTo')}`,
                          },
                        )}
                        <br />
                        {translate('right_menu.shipping_options.round_cost', {
                          deliveryFee: get(
                            standard,
                            'nextRound.slot.cost',
                            '-',
                          ),
                        })}
                      </ShippingMethodDescriptionText>
                    </ShippingMethodDescription>
                  </ShippingMethod>
                )}
              </SectionContent>
            )}

          {/* Delivery Time Slot */}
          <SectionContent>
            {(!isEmpty(intervals) || !isEmpty(mobileIntervals)) &&
            !clearShipping ? (
              <TimeSlotContainer
                shippingMethod={
                  deliveryMethod === 'click_and_collect'
                    ? 'tops'
                    : shippingMethod
                }
                shippingMethods={shippingMethods}
                storeLocator={storeLocator}
                collector={collector}
                slotId={currentSlotId}
                intervals={intervals}
                mobileIntervals={mobileIntervals}
                onChooseSlot={onChooseSlot}
                deliveryMethod={deliveryMethod}
                cart={cart}
              />
            ) : !isEmpty(shippingAddress) && isEmpty(intervals) ? (
              <NoSlotAvalible>
                <TextGuide type="body" color="red">
                  {translate(
                    'checkout_delivery.shipping_method.not_found_slot',
                  )}
                </TextGuide>
              </NoSlotAvalible>
            ) : (
              <NoSlotAvalible>
                <TextGuide type="body" color="red">
                  {deliveryMethod === 'click_and_collect'
                    ? translate(
                        'checkout_delivery.shipping_method.no_select_pickup',
                      )
                    : translate(
                        'checkout_delivery.shipping_method.no_select_shipping',
                      )}
                </TextGuide>
              </NoSlotAvalible>
            )}
          </SectionContent>
        </Section>
      </React.Fragment>
    );
  }
}
const mapDispatchToProps = dispatch => ({});

const mapStateToProps = state => ({
  lang: get(
    find(state.locale.languages, lang => lang.active === true),
    'code',
    'th_TH',
  ),
  clearShipping: state.customer.clearShipping,
});

export default withLocales(connect(mapStateToProps)(ShippingMethodContainer));
