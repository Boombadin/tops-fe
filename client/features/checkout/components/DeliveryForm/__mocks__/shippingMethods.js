export const mockShippingMethods = [
  {
    carrierCode: 'tops',
    methodCode: 'express',
    carrierTitle: 'จัดส่งด่วน',
    methodTitle: 'จัดส่งด่วน',
    amount: 150,
    baseAmount: 150,
    available: true,
    extensionAttributes: {
      deliverySlot: [
        {
          date: '2020-11-17',
          isAllow: undefined,
          slot: [
            {
              available: 200,
              cost: 80,
              enabled: false,
              id: 20,
              isAllow: false,
              quota: 200,
              round: undefined,
              timeFrom: '10:00',
              timeTo: '11:00',
            },
            {
              available: 200,
              cost: 80,
              enabled: false,
              id: 21,
              isAllow: false,
              quota: 200,
              round: undefined,
              timeFrom: '11:00',
              timeTo: '12:00',
            },
            {
              available: 199,
              cost: 100,
              enabled: true,
              id: 22,
              isAllow: true,
              quota: 200,
              round: undefined,
              timeFrom: '12:00',
              timeTo: '13:00',
            },
          ],
        },
      ],
      shipping_description: [
        {
          key: 'caption',
          value:
            'รับสินค้าเร็วขึ้นด้วยการจัดส่งแบบด่วน หรือจัดส่งแบบปกติภายในวันเมื่อสั่งก่อนบ่าย',
        },
        {
          key: 'slot_label',
          value: 'รับสินค้าเร็วขึ้นด้วยบริการจัดส่งแบบด่วน',
        },
      ],
    },
    priceExclTax: 150,
    priceInclTax: 150,
  },
  {
    carrierCode: 'standard',
    methodCode: 'mds',
    carrierTitle: 'จัดส่งถึงบ้าน',
    methodTitle: 'จัดส่งปกติ',
    amount: 60,
    baseAmount: 60,
    available: true,
    extensionAttributes: {
      shipping_description: [
        {
          key: 'caption',
          value:
            'รับสินค้าเร็วขึ้นด้วยการจัดส่งแบบด่วน หรือจัดส่งแบบปกติภายในวันเมื่อสั่งก่อนบ่าย',
        },
        {
          key: 'slot_label',
          value: 'เลือกช่วงเวลารับสินค้าได้ 7 วันล่วงหน้า',
        },
      ],
    },
    priceExclTax: 60,
    priceInclTax: 60,
  },
  {
    carrierCode: 'pickupatstore',
    methodCode: 'tops',
    carrierTitle: 'รับสินค้าที่สาขา / ไดร์ฟ-ทรู',
    methodTitle: 'รับสินค้าที่สาขา / ไดร์ฟ-ทรู',
    amount: 0,
    baseAmount: 0,
    available: true,
    extensionAttributes: {
      shipping_description: [
        {
          key: 'caption',
          value:
            'ช็อปออนไลน์ง่ายๆ และรับสินค้าที่สาขาใกล้บ้าน ฟรีค่าบริการ! ไม่มีขั้นต่ำ',
        },
        {
          key: 'slot_label',
          value: 'ฟรีค่าบริการ! เลือกช่วงเวลารับสินค้าได้ 7 วันล่วงหน้า',
        },
      ],
      store_pickup_enabled: true,
      allowed_retailers: [],
    },
    priceExclTax: 0,
    priceInclTax: 0,
  },
];
