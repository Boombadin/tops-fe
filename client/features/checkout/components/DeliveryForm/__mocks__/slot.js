export const mockShippingMethodSelected = [
  {
    date: '2020-11-17',
    isAllow: undefined,
    slot: [
      {
        available: 200,
        cost: 80,
        enabled: false,
        id: 20,
        isAllow: false,
        quota: 200,
        round: undefined,
        timeFrom: '10:00',
        timeTo: '11:00',
      },
      {
        available: 200,
        cost: 80,
        enabled: false,
        id: 21,
        isAllow: false,
        quota: 200,
        round: undefined,
        timeFrom: '11:00',
        timeTo: '12:00',
      },
      {
        available: 199,
        cost: 100,
        enabled: true,
        id: 22,
        isAllow: true,
        quota: 200,
        round: undefined,
        timeFrom: '12:00',
        timeTo: '13:00',
      },
    ],
  },
];
