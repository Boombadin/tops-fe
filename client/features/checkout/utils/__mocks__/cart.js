export const cartBundleItemsTypeBOGO = [
  {
    id: 386330,
    name: '5200039300',
    description: 'bogo',
    qty_step: 2,
    action_discount_type: 'central_eachn_percent_discount',
    discount_amount: 100,
    items: [
      {
        item_id: 11370572,
        sku: '8852043004598',
        qty: 1,
      },
      {
        item_id: 11370572,
        sku: '8852043004598',
        qty: 1,
      },
    ],
  },
];

export const cartBundleItemsTypeBXGV = [
  {
    id: 385923,
    name: '"2300000755"',
    description: 'b1gv',
    qty_step: 2,
    action_discount_type: '"groupn"',
    discount_amount: 1100,
    items: [
      {
        item_id: 11371187,
        sku: '4901301230591',
        qty: 1,
      },
      {
        item_id: 11371187,
        sku: '4901301230591',
        qty: 1,
      },
    ],
  },
];

export const cartBundleItemsTypeBOGOAndBXGV = [
  {
    id: 386330,
    name: '5200039300',
    description: 'bogo',
    qty_step: 2,
    action_discount_type: 'central_eachn_percent_discount',
    discount_amount: 100,
    items: [
      {
        item_id: 11370572,
        sku: '8852043004598',
        qty: 1,
      },
      {
        item_id: 11370572,
        sku: '8852043004598',
        qty: 1,
      },
    ],
  },
  {
    id: 385923,
    name: '"2300000755"',
    description: 'b1gv',
    qty_step: 2,
    action_discount_type: '"groupn"',
    discount_amount: 1100,
    items: [
      {
        item_id: 11371187,
        sku: '4901301230591',
        qty: 1,
      },
      {
        item_id: 11371187,
        sku: '4901301230591',
        qty: 1,
      },
    ],
  },
];


