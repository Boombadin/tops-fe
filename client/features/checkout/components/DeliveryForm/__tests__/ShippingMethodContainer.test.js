import { fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '@client/config/theme';
import ShippingMethodContainer from '@client/features/checkout/components/DeliveryForm/ShippingMethodContainer';

import { mockShippingMethods } from '../__mocks__/shippingMethods';
import { mockShippingMethodSelected } from '../__mocks__/slot';

jest.mock('@client/hoc/withLocales', () => {
  return Component => props => (
    <Component translate={jest.fn(key => key)} {...props} />
  );
});

jest.mock('react-redux', () => ({
  connect: jest.fn(() => ReactComponent => props => (
    <ReactComponent {...props} />
  )),
}));

jest.mock('react-localize-redux', () => ({
  getTranslate: jest.fn(),
}));

jest.mock('../TimeSlotContainer.js', () => () => <mock-widget />);

jest.mock('@client/components/TimeSlotTab/utils', () => ({
  createIntervals: jest.fn().mockReturnValue(jest.fn()),
}));

jest.mock('@client/features/checkout/utils', () => ({
  getShippingMethod: jest.fn().mockReturnValue({
    method: {
      amount: 150,
      available: true,
      baseAmount: 150,
      carrierCode: 'tops',
      carrierTitle: 'จัดส่งด่วน',
      errorMessage: '',
      extensionAttributes: {
        date: '2020-11-17',
        isAllow: undefined,
        slot: [
          {
            available: 200,
            cost: 80,
            enabled: false,
            id: 20,
            isAllow: false,
            quota: 200,
            round: undefined,
            timeFrom: '10:00',
            timeTo: '11:00',
          },
          {
            available: 200,
            cost: 80,
            enabled: false,
            id: 21,
            isAllow: false,
            quota: 200,
            round: undefined,
            timeFrom: '11:00',
            timeTo: '12:00',
          },
          {
            available: 199,
            cost: 100,
            enabled: true,
            id: 22,
            isAllow: true,
            quota: 200,
            round: undefined,
            timeFrom: '12:00',
            timeTo: '13:00',
          },
        ],
      },
      methodCode: 'express',
      methodTitle: 'จัดส่งด่วน',
      priceExclTax: 150,
      priceInclTax: 150,
    },
    nextRound: {
      date: '2020-11-17',
      slot: {
        available: 199,
        cost: 100,
        enabled: true,
        id: 22,
        isAllow: true,
        quota: 200,
        round: undefined,
        timeFrom: '12:00',
        timeTo: '13:00',
      },
    },
    slotLabel: 'รับสินค้าเร็วขึ้นด้วยบริการจัดส่งแบบด่วน',
  }),
}));
const TEST_ID = {
  SELECT_SHIPPING_METHOD_STANDARD:
    'btn-changeShippingMethodContainer-SelectShippingStandard',
  SELECT_SHIPPING_METHOD_EXPRESS:
    'btn-changeShippingMethodContainer-SelectShippingExpress',
  SHIPPING_TITLE: 'inf-viewShippingMethodContainer-ShippingTitle',
};

describe('features/checkout/components/DeliveryForm/ShippingMethodContainer', () => {
  const initialState = {
    deliveryMethod: 'home_delivery',
    shippingMethods: mockShippingMethods,
    shippingMethod: 'express',
    clearShipping: false,
    intervals: jest.fn(),
    mobileIntervals: jest.fn(),
    shippingMethodSelected: mockShippingMethodSelected,
    active: true,
    shippingAddress: jest.fn(),
    billingAddress: jest.fn(),
    storeLocator: jest.fn(),
    storeLocatorSlot: jest.fn(),
    onChooseSlot: jest.fn(),
    onShippingChange: jest.fn(),
    collector: undefined,
    cart: jest.fn(),
    checked: true,
    onClick: jest.fn(),
  };
  test('Delivery method is Home Delivery | it should show title is home delivery', () => {
    const { asFragment, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ShippingMethodContainer {...initialState} />
      </ThemeProvider>,
    );
    expect(getByTestId(TEST_ID.SHIPPING_TITLE)).toBeInTheDocument();
    expect(
      screen.getByText('checkout_delivery.shipping_method.home_delivery.title'),
    ).toBeInTheDocument();
    expect(
      asFragment(<ShippingMethodContainer {...initialState} />),
    ).toMatchSnapshot();
  });

  test('Delivery method is Click and Collect | it should show title is Click and Collect', () => {
    const initialState = {
      deliveryMethod: 'click_and_collect',
      collector: jest.fn(),
    };
    const { asFragment, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ShippingMethodContainer {...initialState} />
      </ThemeProvider>,
    );
    expect(getByTestId(TEST_ID.SHIPPING_TITLE)).toBeInTheDocument();
    expect(
      screen.getByText(
        'checkout_delivery.shipping_method.click_and_collect.title',
      ),
    ).toBeInTheDocument();
    expect(
      asFragment(<ShippingMethodContainer {...initialState} />),
    ).toMatchSnapshot();
  });
  test('Delivery method is Home Delivery | it should set default shipping method is Express ', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ShippingMethodContainer {...initialState} />
      </ThemeProvider>,
    );
    expect(
      getByTestId(TEST_ID.SELECT_SHIPPING_METHOD_EXPRESS),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        'checkout_delivery.shipping_method.shipping_title_method.express',
      ),
    ).toBeInTheDocument();
  });
  test('Delivery method is Home Delivery and Express is not available  | it should set default shipping method is Standard ', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ShippingMethodContainer {...initialState} />
      </ThemeProvider>,
    );
    expect(
      getByTestId(TEST_ID.SELECT_SHIPPING_METHOD_STANDARD),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        'checkout_delivery.shipping_method.shipping_title_method.delivery',
      ),
    ).toBeInTheDocument();
  });
  test('Click change Shipping method | it should be call function handleShippingMethod', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ShippingMethodContainer {...initialState} />
      </ThemeProvider>,
    );
    const selectStandard = getByTestId(TEST_ID.SELECT_SHIPPING_METHOD_STANDARD);
    fireEvent.click(selectStandard);
    expect(selectStandard).toBeInTheDocument();
    const selectExpress = getByTestId(TEST_ID.SELECT_SHIPPING_METHOD_EXPRESS);
    fireEvent.click(selectExpress);
    expect(selectExpress).toBeInTheDocument();
  });
});
