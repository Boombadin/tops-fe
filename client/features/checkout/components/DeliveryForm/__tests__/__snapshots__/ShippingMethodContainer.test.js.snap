// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/checkout/components/DeliveryForm/ShippingMethodContainer Delivery method is Click and Collect | it should show title is Click and Collect 1`] = `
<DocumentFragment>
  <div
    class="ShippingMethodContainer__Section-e4v8n2-0 gdXndw"
  >
    <div
      class="ShippingMethodContainer__SectionTitle-e4v8n2-1 cEppTm"
      data-testid="inf-viewShippingMethodContainer-ShippingTitle"
    >
      <p
        class="Text-sc-9p67zt-0 kuopgj"
        color="default"
        display=""
        size="15"
        spacing="-0.6px"
      >
        <img
          class="ShippingMethodContainer__SectionTitleIcon-e4v8n2-2 fhDISq"
          src="/assets/icons/calendar-icon.svg"
          style="margin-right: 10px;"
        />
        checkout_delivery.shipping_method.click_and_collect.title
      </p>
    </div>
    <div
      class="ShippingMethodContainer__SectionContent-e4v8n2-3 clFyYt"
    >
      <mock-widget />
    </div>
  </div>
</DocumentFragment>
`;

exports[`features/checkout/components/DeliveryForm/ShippingMethodContainer Delivery method is Home Delivery | it should show title is home delivery 1`] = `
<DocumentFragment>
  <div
    class="ShippingMethodContainer__Section-e4v8n2-0 gdXndw"
  >
    <div
      class="ShippingMethodContainer__SectionTitle-e4v8n2-1 cEppTm"
      data-testid="inf-viewShippingMethodContainer-ShippingTitle"
    >
      <p
        class="Text-sc-9p67zt-0 kuopgj"
        color="default"
        display=""
        size="15"
        spacing="-0.6px"
      >
        <img
          class="ShippingMethodContainer__SectionTitleIcon-e4v8n2-2 fhDISq"
          src="/assets/icons/calendar-icon.svg"
          style="margin-right: 10px;"
        />
        checkout_delivery.shipping_method.home_delivery.title
      </p>
    </div>
    <div
      class="ShippingMethodContainer__SectionContent-e4v8n2-3 kKwRB"
    >
      <div
        class="ShippingMethodContainer__ShippingMethod-e4v8n2-4 kpJTsW"
        data-testid="btn-changeShippingMethodContainer-SelectShippingExpress"
      >
        <label
          class="radio-container ButtonRadio__ButtonRadioWrapper-sc-1c2fqmv-0 jmXno"
        >
          <input
            name="shippingMethod"
            readonly=""
            type="radio"
          />
          <div
            class="radio-checkmark ButtonRadio__RadioCheckMark-sc-1c2fqmv-1 dRrEo"
          />
        </label>
        <div
          class="ShippingMethodContainer__ShippingMethodDescription-e4v8n2-5 kDrWlS"
        >
          <div
            class="ShippingMethodContainer__ShippingMethodDescriptionTitle-e4v8n2-6 ibPjvO"
          >
            <span
              class="ShippingMethodContainer__MethodTitle-e4v8n2-7 bLqiQm"
            >
              checkout_delivery.shipping_method.shipping_title_method.express
            </span>
            <div
              class="ShippingMethodContainer__MethodPowered-e4v8n2-8 gszani"
            >
              <p
                class="Text-sc-9p67zt-0 ShippingMethodContainer__ShippingMethodDescriptionTitleLabel-e4v8n2-9 gQzGIY TextGuide__Caption-sc-1wuo3jj-0 DAZVd"
                color="default"
                display=""
                size="11"
                spacing="normal"
              >
                right_menu.shipping_options.powered_by
              </p>
              <img
                class="ShippingMethodContainer__GrabIcon-e4v8n2-11 cUbjLu"
                src="/assets/images/Grab_logo.svg"
              />
            </div>
          </div>
          <p
            class="Text-sc-9p67zt-0 default-font TextGuide__Caption-sc-1wuo3jj-0 khELOm"
            color="#666666"
            display=""
            size="11"
            spacing="normal"
          >
            รับสินค้าเร็วขึ้นด้วยบริการจัดส่งแบบด่วน
            <br />
            right_menu.shipping_options.next_round right_menu.shipping_options.next_round_datetime
            <br />
            right_menu.shipping_options.round_cost
          </p>
        </div>
      </div>
      <div
        class="ShippingMethodContainer__ShippingMethod-e4v8n2-4 kpJTsW"
        data-testid="btn-changeShippingMethodContainer-SelectShippingStandard"
      >
        <label
          class="radio-container ButtonRadio__ButtonRadioWrapper-sc-1c2fqmv-0 jmXno"
        >
          <input
            name="shippingMethod"
            readonly=""
            type="radio"
          />
          <div
            class="radio-checkmark ButtonRadio__RadioCheckMark-sc-1c2fqmv-1 dRrEo"
          />
        </label>
        <div
          class="ShippingMethodContainer__ShippingMethodDescription-e4v8n2-5 kDrWlS"
        >
          <div
            class="ShippingMethodContainer__ShippingMethodDescriptionTitle-e4v8n2-6 ibPjvO"
          >
            checkout_delivery.shipping_method.shipping_title_method.delivery
          </div>
          <p
            class="Text-sc-9p67zt-0 ShippingMethodContainer__ShippingMethodDescriptionText-e4v8n2-10 dNYNbm TextGuide__Caption-sc-1wuo3jj-0 DAZVd"
            color="default"
            display=""
            size="11"
            spacing="normal"
          >
            รับสินค้าเร็วขึ้นด้วยบริการจัดส่งแบบด่วน
            <br />
            right_menu.shipping_options.next_round right_menu.shipping_options.next_round_datetime
            <br />
            right_menu.shipping_options.round_cost
          </p>
        </div>
      </div>
    </div>
    <div
      class="ShippingMethodContainer__SectionContent-e4v8n2-3 clFyYt"
    >
      <mock-widget />
    </div>
  </div>
</DocumentFragment>
`;
